src/networking.c

```diff
@@ -2116,6 +2117,48 @@ void resetClient(client *c) {
     }
 }
 
+/* Initializes the shared query buffer to a new sds with the default capacity */
+void initSharedQueryBuf(void) {
+    shared_qb = sdsnewlen(NULL, PROTO_IOBUF_LEN);
+    sdsclear(shared_qb);
+}
+
+/* Resets the client's query buffer.
+ *
+ * If the client is using the shared query buffer, the remaining data is copied
+ * to a new private buffer. If the client is using a private buffer, the buffer
+ * is trimmed to the current position. */
+void resetClientQueryBuffer(client *c) {
+    if (c->querybuf == NULL) {
+        return;
+    }
+
+    serverAssert(c->qb_pos <= sdslen(c->querybuf));
+
+    if (c->querybuf != shared_qb) {
+        if (c->qb_pos > 0) {
+            sdsrange(c->querybuf, c->qb_pos, -1);
+            c->qb_pos = 0;
+        }
+        return;
+    }
+
+    size_t remaining = sdslen(c->querybuf) - c->qb_pos;
+
+    if (remaining > 0) {
+        /* Allocate large enough buf to avoid re-allocation in the next read */
+        size_t size = remaining <= PROTO_IOBUF_LEN ? PROTO_IOBUF_LEN : remaining;
+        c->querybuf = sdsnewlen(NULL, size);
+        memcpy(c->querybuf, shared_qb + c->qb_pos, remaining);
```

Comment:
Since we need to move the data to the start of the query buffer anyway, the memcpy operation by itself doesn't add any additional cost.

Reply:
I looked at it again, and I think you're right. Since data will remain in buffer mainly with pipeline clients, in case the client issued, for example, 15KB of commands, we will copy it in this function. however no trimming will be done If we do execute all the commands and no data is left in the buffer - changed accordingly to use the shared qb  and avoid calling `memcpy`.